README.md
# Red Machete Website

> Finest Rock 'N' Roll from Austria: [www.redmachete.at](www.redmachete.at)

## Setup

```bash
nvm use 10
npm install
npm run dev
```

## Build

```bash
npm run build
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.5.1](https://github.com/klausfischer/redmachete/compare/v1.5.0...v1.5.1) (2021-05-16)


### Bug Fixes

* Correct instagram channel and presskit ([e7c7934](https://github.com/klausfischer/redmachete/commit/e7c79348963872b9095386df3906fd2f6db0a047))

## [1.5.0](https://github.com/klausfischer/redmachete/compare/v1.4.0...v1.5.0) (2020-09-06)


### Features

* Add gigs for 2020 ([2e1306d](https://github.com/klausfischer/redmachete/commit/2e1306d94aee950d01fb3e142d9a6d153fb5b656))

## [1.4.0](https://github.com/klausfischer/redmachete/compare/v1.3.1...v1.4.0) (2019-11-17)


### Features

* Add new dates and presspics ([55cb909](https://github.com/klausfischer/redmachete/commit/55cb90921696ff583345d8a66d10c0064ce76078))
* Lazyload header image, add dates anchor ([132f40a](https://github.com/klausfischer/redmachete/commit/132f40aa5b5d23c33953b46707d288e8b7ac5320))
* Replace presspics and presstext ([2a066f3](https://github.com/klausfischer/redmachete/commit/2a066f3875b41843b0129bfb0f6631f8ac13ccc8))


### Bug Fixes

* Remove background color on mobile index ([9c8876e](https://github.com/klausfischer/redmachete/commit/9c8876e8ea89ea3d973248122209b51e3ff999a6))
* Reposition mobile images ([5257722](https://github.com/klausfischer/redmachete/commit/5257722b309c7b1f01fe41bd624250b2587dd221))
* Update dependencies, remove @babel/register ([d8a630c](https://github.com/klausfischer/redmachete/commit/d8a630c26f6991e016b9fb2470001873f09978ea))
* Update merch availability and mobile index image ([b3718c0](https://github.com/klausfischer/redmachete/commit/b3718c0aea273570e057868eb728f44bc5a4f011))
* Update share image ([bf305be](https://github.com/klausfischer/redmachete/commit/bf305beb3fd39547cc787ddc5c3c6f7a4241aa27))

<a name="1.3.1"></a>
## [1.3.1](https://github.com/klausfischer/redmachete/compare/v1.3.0...v1.3.1) (2019-05-23)


### Bug Fixes

* Add 2019 dates to index page ([2ec13fa](https://github.com/klausfischer/redmachete/commit/2ec13fa))



<a name="1.3.0"></a>
# [1.3.0](https://github.com/klausfischer/redmachete/compare/v1.2.0...v1.3.0) (2019-05-23)


### Features

* Add more gigs ([5d75089](https://github.com/klausfischer/redmachete/commit/5d75089))
* Update presskit link ([39547a2](https://github.com/klausfischer/redmachete/commit/39547a2))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/klausfischer/redmachete/compare/v1.1.1...v1.2.0) (2019-01-02)


### Features

* Add fb links ([bcf31e6](https://github.com/klausfischer/redmachete/commit/bcf31e6))
* Add more gigs ([38e4857](https://github.com/klausfischer/redmachete/commit/38e4857))



<a name="1.1.1"></a>
## [1.1.1](/compare/v1.1.0...v1.1.1) (2018-04-14)


### Bug Fixes

* Also ship sitemap.xml aca5ddd
* Update landing image 69dc9d7



<a name="1.1.0"></a>
# 1.1.0 (2018-04-14)


### Bug Fixes

* Add invisible text for links df8b8b0
* Increase gigtable mobile font size a93c378
* **pwa:** Add name to manifest.json 9630ed8


### Features

* Add GA 33dafa6
* Add shitty Videocontroller y’all! 7a8b45f
* Add sitemap.xml ac49944
* Enable object-spread for js ddc1e6d


gulpfile.babel.js
import autoprefixer from 'autoprefixer';
import browserSync from 'browser-sync';
import cssnano from 'cssnano';
import del from 'del';
import fs from 'fs';
import gulp from 'gulp';
import imagemin from 'gulp-imagemin';
import nunjucksRender from 'gulp-nunjucks-render';
import php from 'gulp-connect-php';
import pngquant from 'imagemin-pngquant';
import postCSS from 'gulp-postcss';
import pump from 'pump';
import sass from 'gulp-sass';
import sassGlob from 'gulp-sass-glob';
import config from './config.json';
import data from './data';
import pkg from './package.json';

const svgoPlugins = [
  { cleanupIDs: false },
  { cleanupNumbericValues: false },
  { cleanupListOfValues: false },
  { collapseGroups: false },
  { convertStyleToAttrs: true },
  { mergePaths: false },
  { removeComments: true },
  { removeStyleElement: true },
  { removeDoctype: true },
];

let d = {}; // JSONs from data folder

gulp.task('clean', () => {
  del.sync('dist');
});

gulp.task('copy', () =>
  gulp.src(['.htaccess', 'sitemap.xml', 'vinyldownload/**/*'], { base: '.' })
    .pipe(gulp.dest(config.dest.root)));

gulp.task('loadData', (cb) => {
  data(config.src.data).then((res) => {
    d = res;
    cb();
  });
});

gulp.task('sass', () =>
  gulp.src('./scss/main.scss')
    .pipe(sassGlob())
    .pipe(sass({ outputStyle: 'expanded' }).on('error', sass.logError))
    .pipe(postCSS([
      autoprefixer({ browsers: pkg.browserslist }),
      cssnano(),
    ]))
    .pipe(gulp.dest(config.dest.styles))
    .pipe(browserSync.stream()));

function extendNunjucksEnv(environment) {
  environment.addFilter('includeFile', path => fs.readFileSync(`./${path}`));
}


gulp.task('nunjucks', ['loadData', 'images'], (cb) => {
  pump([
    gulp.src(config.src.pages),
    nunjucksRender({
      data: d,
      path: `./${config.src.templates}/`,
      manageEnv: extendNunjucksEnv,
    }),
    gulp.dest(config.dest.root),
  ], cb);
});

gulp.task('images', () =>
  gulp.src(config.src.images)
    .pipe(imagemin({
      progressive: true,
      svgoPlugins,
      use: [pngquant()],
    }))
    .pipe(gulp.dest(config.dest.images)));

gulp.task('watch', () => {
  gulp.watch(`${config.src.styles}/**/*.scss`, ['sass']);
  gulp.watch(['.htaccess', 'vinyldownload/**/*'], ['copy', browserSync.reload]);
  gulp.watch([config.src.pages, 'data/**/*.json', `${config.src.templates}/**/*.html`], ['nunjucks', browserSync.reload]);
});

gulp.task('php', () => {
  php.server({
    hostname: '0.0.0.0',
    base: 'dist',
    port: 4040,
    open: false,
  });
});

gulp.task('browser-sync', ['php'], () => {
  browserSync.init({
    open: true,
    port: 4200,
    proxy: {
      target: 'http://localhost:4040',
      reqHeaders() {
        return {
          'accept-encoding': 'identity',
          agent: false,
        };
      },
    },
    reloadDelay: 1000,
  });
});

gulp.task('default', ['clean', 'copy', 'sass', 'nunjucks', 'images', 'watch', 'browser-sync']);
gulp.task('build',   ['clean', 'copy', 'sass', 'nunjucks', 'images']);


webpack.config.js
const path = require('path');
const eslintFriendlyFormatter = require('eslint-friendly-formatter');
const webpack = require('webpack');

function resolve(dir) {
  return path.join(__dirname, dir);
}

const config = mode => ({
  entry: {
    redmachete: './js/main.js',
  },
  output: {
    path: resolve('dist/scripts'),
    filename: '[name].js',
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        include: resolve('js'),
        enforce: 'pre',
        use: {
          loader: 'eslint-loader',
          options: {
            formatter: eslintFriendlyFormatter,
          },
        },
      },
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
        },
      },
    ],
  },
  mode,
  plugins: [
    new webpack.LoaderOptionsPlugin({ options: {} }),
  ],
});

module.exports = config;


js/VideoController.js
import util from './util';

const VideoController = (videoTrigger, opts) => {
  const defaults = {
    youtubeUrl: '',
    videoContainer: '#videoOverlay',
  };
  const options = { ...defaults, ...opts };
  options.template = `<div style="max-width: 80%;" class="mt3 center"><div class="aspect-ratio aspect-ratio--16x9"><iframe class="aspect-ratio--object" frameborder="0" src="${options.youtubeUrl}" allowfullscreen></iframe></div></div>`;

  const trigger = videoTrigger;
  const vcontainer = document.querySelector(options.videoContainer);

  const hideVideo = () => {
    vcontainer.classList.add('hidden');
    // Reset iframe URL to stop video
    const iframe = vcontainer.querySelector('iframe');
    iframe.setAttribute('src', iframe.getAttribute('src'));
    vcontainer.removeEventListener('click', hideVideo);
  };

  const showVideo = () => {
    // Append iframe with youtube video
    util.emptyNode(options.videoContainer);
    const iframe = util.htmlToElements(options.template);
    vcontainer.appendChild(iframe[0]);
    // Show container and enable overlay closing
    vcontainer.classList.remove('hidden');
    vcontainer.addEventListener('click', hideVideo);
  };

  const init = () => {
    trigger.addEventListener('click', showVideo);
  };

  return {
    init,
  };
};

export default VideoController;


js/VinylController.js

const CONFIG = {
  vinylApi: `${window.location.origin}/vinyldownload/api.php`,
  downloadPage: `${window.location.origin}/vinyldownload/download.php`,
};

const ALERT_SELECTOR = '.rm-alert';
const CODE_INPUT_SELECTOR = '#form-vinyl__vinylcode';
let httpRequest;

function showAlert(msg, type) {
  const alerts = document.querySelectorAll(ALERT_SELECTOR);
  for (let i = 0; i < alerts.length; i += 1) {
    alerts[i].classList.add('hidden');
  }

  const activeAlert = document.querySelector(`${ALERT_SELECTOR}--${type}`);
  activeAlert.innerHTML = msg;
  activeAlert.classList.remove('hidden');
}

function showResults() {
  if (httpRequest.readyState === XMLHttpRequest.DONE) {
    switch (httpRequest.status) {
      case 200: {
        showAlert('Thanks! Download starts immediately', 'success');
        const code = document.querySelector(CODE_INPUT_SELECTOR).value;
        // redirect to download.php
        window.location.replace(`${CONFIG.downloadPage}?${code}`);
        break;
      }
      case 400: {
        showAlert('Wrong code length.', 'danger');
        break;
      }
      case 404: {
        showAlert('Please try a correct code.', 'danger');
        break;
      }
      case 410: {
        showAlert('Your code was already redeemed, please contact us via <a class="link dim" href="mailto:band@redmachete.at">mailto:band@redmachete.at</a> and add your code to the email', 'danger');
        break;
      }
      default: {
        showAlert('Couldn\'t establish connection.', 'danger');
        break;
      }
    }
  }
}

function submitVinylCode(e) {
  e.preventDefault();
  httpRequest = new XMLHttpRequest();

  if (!httpRequest) {
    showAlert('Couldn\'t create Request', 'danger');
    return false;
  }
  const code = document.querySelector(CODE_INPUT_SELECTOR).value;

  if (code) {
    httpRequest.onreadystatechange = showResults;
    httpRequest.open('GET', `${CONFIG.vinylApi}?code=${code}`);
    httpRequest.send();
  }
  return true;
}

export default {
  submitVinylCode,
};


gulpfile.js
const autoprefixer = require('autoprefixer');
const browserSync = require('browser-sync');
const cssnano = require('cssnano');
const del = require('del');
const fs = require('fs');
const gulp = require('gulp');
const imagemin = require('gulp-imagemin');
const nunjucksRender = require('gulp-nunjucks-render');
const php = require('gulp-connect-php');
const pngquant = require('imagemin-pngquant');
const postCSS = require('gulp-postcss');
const pump = require('pump');
const sass = require('gulp-sass');
const sassGlob = require('gulp-sass-glob');
const config = require('./config.json');
const data = require('./data');
const pkg = require('./package.json');

const svgoPlugins = [
  { cleanupIDs: false },
  { cleanupNumbericValues: false },
  { cleanupListOfValues: false },
  { collapseGroups: false },
  { convertStyleToAttrs: true },
  { mergePaths: false },
  { removeComments: true },
  { removeStyleElement: true },
  { removeDoctype: true },
];

let d = {}; // JSONs from data folder

gulp.task('clean', () => {
  del.sync('dist');
});

gulp.task('copy', () =>
  gulp.src(['.htaccess', 'sitemap.xml', 'vinyldownload/**/*'], { base: '.' })
    .pipe(gulp.dest(config.dest.root)));

gulp.task('loadData', (cb) => {
  data(config.src.data).then((res) => {
    d = res;
    cb();
  });
});

gulp.task('sass', () =>
  gulp.src('./scss/main.scss')
    .pipe(sassGlob())
    .pipe(sass({ outputStyle: 'expanded' }).on('error', sass.logError))
    .pipe(postCSS([
      autoprefixer({ browsers: pkg.browserslist }),
      cssnano(),
    ]))
    .pipe(gulp.dest(config.dest.styles))
    .pipe(browserSync.stream()));

function extendNunjucksEnv(environment) {
  environment.addFilter('includeFile', path => fs.readFileSync(`./${path}`));
}


gulp.task('nunjucks', ['loadData', 'images'], (cb) => {
  pump([
    gulp.src(config.src.pages),
    nunjucksRender({
      data: d,
      path: `./${config.src.templates}/`,
      manageEnv: extendNunjucksEnv,
    }),
    gulp.dest(config.dest.root),
  ], cb);
});

gulp.task('images', () =>
  gulp.src(config.src.images)
    .pipe(imagemin({
      progressive: true,
      svgoPlugins,
      use: [pngquant()],
    }))
    .pipe(gulp.dest(config.dest.images)));

gulp.task('watch', () => {
  gulp.watch(`${config.src.images}/**/*.jpg`, ['images']);
  gulp.watch(`${config.src.styles}/**/*.scss`, ['sass']);
  gulp.watch(['.htaccess', 'vinyldownload/**/*'], ['copy', browserSync.reload]);
  gulp.watch([config.src.pages, 'data/**/*.json', `${config.src.templates}/**/*.html`], ['nunjucks', browserSync.reload]);
});

gulp.task('php', () => {
  php.server({
    hostname: '0.0.0.0',
    base: 'dist',
    port: 4040,
    open: false,
  });
});

gulp.task('browser-sync', ['php'], () => {
  browserSync.init({
    open: true,
    port: 4200,
    proxy: {
      target: 'http://localhost:4040',
      reqHeaders() {
        return {
          'accept-encoding': 'identity',
          agent: false,
        };
      },
    },
    reloadDelay: 1000,
  });
});

gulp.task('default', ['clean', 'copy', 'sass', 'nunjucks', 'images', 'watch', 'browser-sync']);
gulp.task('build',   ['clean', 'copy', 'sass', 'nunjucks', 'images']);
